test/test-test.js

```diff
@@ -233,6 +233,69 @@
             }).call({}, "arg1", {}, done);
         },
 
+        "async promise-based test": {
+            requiresSupportFor: {
+                Promise: typeof Promise !== "undefined"
+            },
+
+            "restores the sandbox when the promise is fulfilled": function (done) {
+                var method = function () {};
+                var object = { method: method };
+                var spy = sinon.spy();
+                sinon.test(function (callback) {
+                    var stub;
+                    var sandbox = this;
+                    return Promise.resolve()
+                        .then(function () {
```

Comment:
I would greatly simplify these tests. Simply return `Promise.resolve();`. There is no need to test the functionality of the promise library. Remove spies from the equation. We're only testing the sandbox functionality.